test/test_selectors.py

```diff
@@ -0,0 +1,670 @@
+        self.canceled = False
+
+    def cancel(self):
+        self.canceled = True
+
+    def run(self):
+        time.sleep(self.timeout)
+        if not self.canceled:
+            os.kill(os.getpid(), signal.SIGALRM)
+
+
+class AlarmMixin(object):
+    alarm_thread = None
+
+    def set_alarm(self, timeout):
+        if not hasattr(signal, "SIGALRM"):
+            self.skipTest("Platform doesn't have signal.SIGALRM")
+        if AlarmMixin.alarm_thread is None:
+            self.addCleanup(self.cancel_alarm)
+        AlarmMixin.alarm_thread = AlarmThread(timeout)
+        AlarmMixin.alarm_thread.start()
+
+    def cancel_alarm(self):
+        if AlarmMixin.alarm_thread is not None:
+            AlarmMixin.alarm_thread.cancel()
+            AlarmMixin.alarm_thread.join(0.0)
+        AlarmMixin.alarm_thread = None
+
+
+@skipUnless(selectors.HAS_SELECT, "Platform doesn't have a selector")
+class WaitForIOTest(unittest.TestCase, AlarmMixin):
+    """ Tests for the higher level wait_for_* functions. """
+    def make_socketpair(self):
+        rd, wr = socketpair()
+        self.addCleanup(rd.close)
+        self.addCleanup(wr.close)
+        return rd, wr
+
+    def test_selector_error(self):
+        err = selectors.SelectorError(1)
+        self.assertEqual(err.__repr__(), "<SelectorError errno=1>")
+        self.assertEqual(err.__str__(), "<SelectorError errno=1>")
+
+    def test_wait_for_read_single_socket(self):
+        rd, wr = self.make_socketpair()
+        self.assertEqual([], selectors.wait_for_read(rd, timeout=0.001))
+
+    def test_wait_for_read_multiple_socket(self):
+        rd, rd2 = self.make_socketpair()
+        self.assertEqual([], selectors.wait_for_read([rd, rd2], timeout=0.001))
+
+    def test_wait_for_read_empty(self):
+        self.assertEqual([], selectors.wait_for_read([], timeout=0.001))
+
+    def test_wait_for_write_single_socket(self):
+        wr, wr2 = self.make_socketpair()
+        self.assertEqual([wr], selectors.wait_for_write(wr, timeout=0.001))
+
+    def test_wait_for_write_multiple_socket(self):
+        wr, wr2 = self.make_socketpair()
+        result = selectors.wait_for_write([wr, wr2], timeout=0.001)
+        # assertItemsEqual renamed in Python 3.x
+        if hasattr(self, "assertItemsEqual"):
+            self.assertItemsEqual([wr, wr2], result)
+        else:
+            self.assertCountEqual([wr, wr2], result)
+
+    def test_wait_for_write_empty(self):
+        self.assertEqual([], selectors.wait_for_write([], timeout=0.001))
+
+    def test_wait_for_non_list_iterable(self):
+        rd, wr = self.make_socketpair()
+        iterable = {'rd': rd}.values()
+        self.assertEqual([], selectors.wait_for_read(iterable, timeout=0.001))
+
+    def test_wait_timeout(self):
+        rd, wr = self.make_socketpair()
+        t = get_time()
+        selectors.wait_for_read([rd], timeout=0.01)
+        t = get_time() - t
+        self.assertGreater(t, 0.0)
+        self.assertLess(t, 0.1)
```

Comment:
Can we package these asserts up into a helper assert we have in a base class? That'd be super convenient. Maybe even in a helper context manager that *also* manages the calls to `get_time`?